Variable Products: Fix price when “Default form values” is set

asumaran opened the pull request:
Fixes #829 

Here we are fixing the default total amount for "Variable Products". This value is set when page loads and is taken from `wc_stripe_payment_request_params.product.total.amount`.

Previously, it failed because the default attributes were returned without the `attribute_` prefix. Because of this we weren't able to get the right `variation_id` using `$data_store->find_matching_product_variation( $product, $attributes )`. 

On this PR we are adding the prefix manually. I've asked on Slack (p1618615018086800-slack-C0E1AV8T0) if it's there a cleaner way to do it. No response. Although I see no downsides on doing it manually since it's only used for this case.

# Testing

I've tested this for "Variable Products". @ricardo suggested I should also test with "Composite" and "Mixed and Match" product types but both disable the button on the Product's page:

**Composite Products**

Plugin users after [v7.1.4](http://changelogs.s3.amazonaws.com/changelogs/woocommerce-composite-products/changelog.txt) (released last December) will not encounter this error: 

<details>
 <summary>
Plugin is hiding the button on the Product Page using `wc_stripe_hide_payment_request_on_product_page` filter.
</summary>

```php
public static function init() {
  add_filter( 'wc_stripe_hide_payment_request_on_product_page', array( __CLASS__, 'hide_stripe_quickpay' ), 10, 2 );
}

/**
 * Hide Stripe Quick-pay buttons for non-static Composites.
 *
 * @since 7.1.4
 *
 */
public static function hide_stripe_quickpay( $hide_button, $post ) {

  global $product;

  $the_product = $product && is_a( $product, 'WC_Product' ) ? $product : wc_get_product( $post->ID );

  if ( $the_product && $the_product->is_type( 'composite' ) ) {
    $hide_button = true;
  }

  return $hide_button;
}
```

</details>

**Mix and Match Products**

<details>
 <summary>
Plugin is using `wc_stripe_payment_request_supported_types` filter to remove support for `mix-and-match` product type on Stripe.
</summary>

```php
public static function init() {

  // Hide Stripe's payment request buttons on MNNM products.
  add_filter( 'wc_stripe_payment_request_supported_types', array( __CLASS__, 'supported_request_product_types' ) );

}

/**
 * Hide Stripe's instant pay buttons
 *
 * @param   array $types - The product types that can support payment request buttons.
 * @return  array
 */
public static function supported_request_product_types( $types ) {
  $key = array_search( 'mix-and-match', $types );

  if ( $key ) { 
    unset( $types[$key] );
  }
  
  return $types;
}
```

</details>

@v18 what do you suggest here? Should I try to make sure this fix works with Composite and Mix and Match product types even though the plugins itself are not supporting the "Buy now" button on the product page? I've already tested with Composite products and it seems to be working fine but I had to hardcode their plugin to enable the button.

### Testing instructions

First of all, use Chrome create the following attributes and its terms as follows:

| Attribute | Terms |
| ------------- | ------------- |
| color | red, green, blue  |
| size | medium, large, extra large |


1. Create new variable product
2. Click "Attributes" tab 
3. Add "color" attribute. Add "red", "green", and "blue" as values. Click "use for variations"
4. Add "size" attribute. Add "extra large", "large", "medium" as values. . Click "use for variations"
5. Click "Save attributes" button
6. Click "Variations" tab and Click "Go" button
7. Click "Any Color" and pick "red". Click "Any size" and pick "medium". Set Regular price to "10"
8. Click "Go button"
9. Click "Any Color" and pick "green". Click "Any size" and pick "large". Set Regular price to "15"
10. Click "Go button"
11. Click "Any Color" and pick "blue". Click "Any size" and pick "extra large". Set Regular price to "20"
12. Set "Default Form Values:" to "red", "medium"
12. Click "Save changes" then Click "Publish" to publish the product.
13. Go to the product's page.
14. You should see `$10.00` as price.
15. Click on "Buy now" button and you should see `$10.00` as total in the Google Payment dialog.

### Testing list

```
Product Attributes
------------------

size        color  Price
------------------------
medium        red     10
large       green     15
extra large  blue     20
extra large   any     25
extra large   red     30
```

DFV = Default Form Values;

- [x] DFV is "medium" and "red" then price is $10.00
- [x] DFV is "large" and "green" then price is $15.00
- [x] DFV is "extra large" and "blue" then price is $20.00
- [x] DFV is "extra large" and "No default color" then price is not set. Clicking "But now" button will result in an error asking the user to pick a "color".
- [x] **DFV is "extra large" and "red" then price is $25.00** (*)
- [x] DFV is "large" and "green" and URL has `?attribute_pa_size=extra-large`. Selected attributes are "extra large" and "green". Then price is $25.
- [x] **DFV is "extra-large" and "blue" and URL has `?attribute_pa_color=red`. Selected attributes are "extra large" and "red". Then price is $25** (**)
- [x] DFV is "large" and "blue". Since there's no variation for it. No variation will be selected. No price is set. User will be asked to pick a variation.
- [x] DFV is "large" and "blue" and URL has `?attribute_pa_size=extra-large`. Selected attributes are "extra-large" and "blue". Then price is $20.

---

*, ** This case actually should be `$30`. But the issue is not in our end. It seems for [WooCommerce](https://github.com/woocommerce/woocommerce/blob/840849ad838cb49b514878d210d0e0d6495cc14f/includes/data-stores/class-wc-product-data-store-cpt.php#L1093-L1174) the position of the attribute pair is more relevant over the exact match of the DFV. I've [created an issue](https://github.com/woocommerce/woocommerce/issues/29726) reporting this on the `woocommerce` repo

asumaran committed:
Use correct price for a Variable product if “Default form values” is set

asumaran commented:
I've tested this with Apple Pay and it works exactly the same as using Google Payments.

asumaran commented:
Port for `woocommerce-payments` is ready at https://github.com/Automattic/woocommerce-payments/pull/1642

v18 commented:
> what do you suggest here? Should I try to make sure this fix works with Composite and Mix and Match product types even though the plugins itself are not supporting the "Buy now" button on the product page?

> I've already tested with Composite products and it seems to be working fine but I had to hardcode their plugin to enable the button.

Thanks @asumaran! If I understand you correctly, you're saying that the payment request button feature does work with Composite products in this PR - is that right?

I'm not sure this PR alone will allow the payment request button to be used with Mix and Match products. This isn't a blocker for this PR, since this PR is solving a different bug. 

For testing purposes, I agree that it makes sense to hardcode a change in the extension to see that it's working.

If you are able to test the problem that's happening with Mix and Match and Composite products that's great, and you can outline the remaining problems that need to be solved as a comment in #1401.

As for the filters those extensions are using - once we have support for those plugins in our Stripe gateway extension, we can communicate about it to let the authors know so they can remove their filters if they'd like.

dechov commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -358,7 +368,7 @@ public function get_product_data() {
 		$data['total']        = [
 			'label'   => apply_filters( 'wc_stripe_payment_request_total_label', $this->total_label ),
-			'amount'  => WC_Stripe_Helper::get_stripe_amount( $product->get_price() ),
+			'amount'  => $total_amount,
```
Does this change contribute to the fix, or is it more of an unrelated refactor?

I see a similar change [here](https://github.com/Automattic/woocommerce-payments/pull/1642/files#diff-01da8e7c10c31988f0205e9327a25bcaaf64e6485c6d391912b68dc66d808353R254) in the WCPay PR except that a different amount param (item instead of total) is being replaced.

asumaran replied:
Good catch. I'll remove those lines. I've made another approach before. I'll update the code in WCPay if needed

dechov commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -311,16 +311,26 @@ public function get_product_data() {
 		if ( 'variable' === $product->get_type() ) {
 			$attributes = wc_clean( wp_unslash( $_GET ) );
 
+			if ( empty( $attributes ) ) {
```
This appears to rely on attributes being the only GET params – is that correct? If I add `?utm_medium=email` to the URL, then I get the original behavior of ignoring the default attributes.

asumaran replied:
I've looked into this and haven't found a case where `$_GET` will actually be used. The `get_product_data` method is only used on the `scripts` method which is used to expose data to the frontend via `add_action( 'wp_enqueue_scripts', [ $this, 'scripts' ] )`.

I'll see where this was introduced to see if it's really necessary otherwise I'll remove it.

dechov replied:
👍 ICYMI setting a `?attribute_size=large` URL param prefills that attribute value.

asumaran replied:
@dechov The code was introduced a year ago on this PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1149 by @DanReyLop 

It seems to me that previously we were able to select a variation via `GET`? It would be nice some guidance here since I don't see if this is case anymore.

asumaran replied:
> 👍 ICYMI setting a ?attribute_size=large URL param prefills that attribute value.

yeah, I was assuming something like that would be possible. In any case what do you suggest here in order to keep that functionality and the fix I'm doing here?

asumaran replied:
actually the query string is something like this: `?attribute_pa_size=extra-large` in order to work properly.

asumaran replied:
@dechov would this be good to you: Ask if is there at least one `$_GET` variable starting with `attribute_` to use `$_GET` variables otherwise to use the `$product->get_default_attributes()`? Something like:

```php
$attributes = at_least_one('attribute_', $_GET) ? wc_clean( wp_unslash( $_GET ) ) : $product->get_default_attributes();
```

dechov replied:
There is WooCommerce core functionality to pre-fill the attribute selection by passing a query param – not sure about the `pa_` as only `attribute_size=large` is working for me. I don't know if this feature is documented, and I only discovered it by looking at that code (introduced in the PR you linked) which attempts compatibility with it by reading from `$_GET` so that those pre-filled attributes can be used to prepare the payment request total as well.

I think the overall goal here is to match the WC core behavior (i.e. show in the payment request flow the same price that appears for the currently selected variation on the page), which appears to fall back to the default attribute if (and only if) a query parameter isn't passed.

asumaran replied:
> There is WooCommerce core functionality to pre-fill the attribute selection by passing a query param

Yeah, I just discovered that.

> I think the overall goal here is to match the WC core behavior

That's what I'm working now. I'll update my PR shortly

dechov reviewed (changes requested):
Tests well using provided instructions – though the case described in the initial set of instructions ("red" + "medium") behaves correctly in trunk as well (because it's the cheapest variation, if I understand correctly). I primarily used the second listed case ("green" + "large") to validate the fix.

I noticed that the solution is bypassed if any query params are present, including some cases where it would still be desirable like if only a partial set of attributes is provided (or if the query params are unrelated). Is there a way to target the "default params" case more precisely and just supply any that are missing?

Haven't looked into it, but could we perhaps reference the code that renders the correct price on the product page itself, and make sure we're doing something similar (if not reusing the same function)?

By the way, I was able to reproduce the original issue in trunk only partially, though well enough to expose the bug: for me, the shipping methods request kicks off immediately after opening the payment modal, after which the correct value in the form is reflected (perhaps related to https://github.com/woocommerce/woocommerce-gateway-stripe/issues/829#issuecomment-486121098).

Google Pay in Chrome (90.0.4430.85) shows me only a loading indicator followed by the correct value. Apple Pay in Safari (both desktop and mobile) shows the incorrect value as described in the issue, followed by the correct value a second later:

Before shipping methods load | After shipping methods load
-- | --
<img width="530" src="https://user-images.githubusercontent.com/1867547/115873631-265d9a80-a411-11eb-9e56-3d0ba6811a74.png"> | <img width="517" src="https://user-images.githubusercontent.com/1867547/115873626-25c50400-a411-11eb-92f0-9967984e6ed7.png">

asumaran commented:
> Tests well using provided instructions – though the case described in the initial set of instructions ("red" + "medium") behaves correctly in trunk as well 

It works in `trunk` because:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/c239804c627f96fc8b5573a149a09461dcecaf74/includes/payment-methods/class-wc-stripe-payment-request.php#L316-L325

In the code above`$variation_id` will be `0` because `$data_store->find_matching_product_variation()` is not finding a corresponding product variation due to the format of `$attributes` was incorrect. So, `$product` was still the Variable Product instead of the selected Variation Product (The one that matches the "Default Form Values").

So `$product->get_price()` will return the [first post_meta entry for (`_price`)](https://github.com/woocommerce/woocommerce/blob/trunk/includes/data-stores/class-wc-product-data-store-cpt.php#L364) for the Variable Product.


```
mysql> SELECT * FROM `wp_postmeta` WHERE `post_id` = 357 AND meta_key = '_price';
+---------+---------+----------+------------+
| meta_id | post_id | meta_key | meta_value |
+---------+---------+----------+------------+
|   16330 |     357 | _price   | 10         |
|   16331 |     357 | _price   | 15         |
|   16332 |     357 | _price   | 20         |
|   16333 |     357 | _price   | 25         |
|   16334 |     357 | _price   | 30         |
+---------+---------+----------+------------+
5 rows in set (0.00 sec)
```

asumaran commented:
> I noticed that the solution is bypassed if any query params are present, including some cases where it would still be desirable like if only a partial set of attributes is provided (or if the query params are unrelated). Is there a way to target the "default params" case more precisely and just supply any that are missing?

@dechov I think that would be nice to implement. We just need to make sure we are addressing all the possible cases correctly. Here's a table where we can see all the possible cases if we have two attributes with one value each. `size: large` and `color: red`.

```
| $_GET \ Default Form Value | size:large - color:red | size:no default - color:red       | size:large - color:no default       | size:no default - color:no default             |
|----------------------------|------------------------|-----------------------------------|-------------------------------------|------------------------------------------------|
| size=large                 | size:large - color:red | size:large - color:red            | size:large - color:choose an option | size:large - color:choose an option            |
| color=red                  | size:large - color:red | size:choose an option - color:red | size:large - color:red              | size:choose an option - color: red             |
| size=large&color=red       | size:large - color:red | size:large - color:red            | size:large - color:red              | size:large - color:red                         |
| (no attribute via GET)     | size:large - color:red | size:choose an option - color:red | size:large - color:choose an option | size:choose an option - color:choose an option |
```

Basically, `$_GET` attributes (if defined) will take precedence and will replace values defined for any attribute via "Default Form Values". Is that what you meant?

This will take a bit more to implement though. But first we need to make sure if we support passing variation attributes via `$_GET` because I haven't seen a case we do use that. /cc @v18

asumaran commented:
>By the way, I was able to reproduce the original issue in trunk only partially, though well enough to expose the bug: for me, the shipping methods request kicks off immediately after opening the payment modal, after which the correct value in the form is reflected (perhaps related to #829 (comment)).

@dechov Yeah, in `trunk` the correct amount is returned after requesting `?wc-ajax=wc_stripe_get_shipping_options`  but it fails for "Composite" and "Mix and Match" product types. That will be handled in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1401

 The issue I'm fixing here is only for the case when the modal is taking the amount from JS via `wc_stripe_payment_request_params.product.total` and it happens only when the page loads.

asumaran commented:
> Google Pay in Chrome (90.0.4430.85) shows me only a loading indicator followed by the correct value. Apple Pay in Safari (both desktop and mobile) shows the incorrect value as described in the issue, followed by the correct value a second later:

<img width="486" alt="Screen Shot 2021-04-23 at 11 24 51 PM" src="https://user-images.githubusercontent.com/1025173/115947122-32863e00-a48b-11eb-89fa-6fd15e1950d8.png">

I can reproduce it in both Safari and Chrome. I think you can avoid the extra request when you have multiple addresses.

asumaran commented:
~https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1533#issuecomment-826030935 BTW I'm implementing this so we support the existing way of passing attributes via GET and via Default Form Values.~

dechov commented:
> I think you can avoid the extra request when you have multiple addresses.

@asumaran The extra request I was referring to is that `wc_stripe_get_shipping_options` call you mentioned – it just kicks off automatically without waiting for an address (presumably because I already have saved addresses) – and in Chrome never shows the original total passed to the payment flow.

asumaran committed:
Use variation value passed via GET otherwise get default value for given attribute

asumaran commented:
@dechov I've updated the PR. I was slightly going off on a tangent. I was trying to validate and also select the right value/attribute for each dropdown but I noticed woocommerce is already doing that. If the variation is not found dropdown is reset.

https://github.com/woocommerce/woocommerce/blob/ef61092ebdef2887a95c87cb36488fefcb92e4c1/assets/js/frontend/add-to-cart-variation.js#L206-L223

Here they use the passed attribute via GET. If not attribute is pased via GET default value is taken.

https://github.com/woocommerce/woocommerce/blob/trunk/includes/wc-template-functions.php#L3013

^ I'm doing exactly the same on this PR. So the price corresponds to the selected variation. Also added more test cases to the PR's description.

dechov commented:
@asumaran Wrote the below at the beginning of the week and didn't realize I hadn't posted it – apologies! I like the recent commit and will test & review as is, since I think this can be opened as a different issue:

> I ran into a related issue in which attribute selections can persist in the browser after a page reload but the payment request flow doesn't originally reflect them. Chrome and Safari aren't keeping all form values for me after refresh (the way Firefox is), but the form values _are_ still there after clicking "Add to cart": if you select "red" + "medium", add to cart, and then open the payment request flow, the selection is initially ignored (instead using default attributes "green" + "large" if using this branch, see [screenshot](https://cloudup.com/cIjd7Qfaha2)).
>
> Perhaps it's out of scope of this PR, but it does seem indicative of a deeper issue with the variation match being done in the front-end but the payment request flow not staying in sync in all cases. Would a solution in JS be worth considering? I see the payment request is updated when the variation is explicitly changed [[code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/c239804c627f96fc8b5573a149a09461dcecaf74/assets/js/stripe-payment-request.js#L613)], but maybe we can also trigger it on some init event soon after page load?

dechov commented:
@asumaran The current approach looks good to me but the implementation isn’t working for me as is, I think because the display name of the attribute (“Color”) is normalized in some additional way when determining the key (“attribute_color”). The latter works as a query param to pre-fill the selector for me – as does leaving it blank – but in neither case does it prepare the payment request data properly, I believe due to the capitalization.

asumaran commented:
@dechov I've tried with capitalized attribute names and everything seems to be working as expected.

<img width="721" alt="Screen Shot 2021-05-03 at 1 54 52 PM" src="https://user-images.githubusercontent.com/1025173/116919614-23d02180-ac17-11eb-80b8-45b8aec94fe9.png">

Do you mind sharing a bit more how are you testing this PR? What attributes do you have set up?

dechov commented:
@asumaran Ah, seems to only happen when adding custom product attributes – on the product page, not Attributes screen – with names that have capital letters or spaces (etc?). Didn't do extensive testing but seems fixable by sanitizing the name when constructing the key (as is done [here](https://github.com/woocommerce/woocommerce/blob/b19500728b4b292562afb65eb3a0c0f50d5859de/includes/data-stores/class-wc-product-data-store-cpt.php#L1103)).

asumaran committed:
Sanitize passed attribute via GET

asumaran commented:
@dechov I think I was able to reproduce the issue you were having. Although, the problem for me was when using a custom attribute **value** with capital letters and not for the attribute name. e.g. adding a `Green` value for the "size" attribute. I've fixed it by sanitizing the value passed via GET.

If you still have the same issue please share more details on how to reproduce it.

dechov commented:
On the product page, when I added to a product a custom product attribute with an "unsanitized" name like "Type" (or "product type") and set a default form value for it, the initial payment request total still didn't reflect the variation with that default – it just falls back to the first variation. Can outline more specific testing instructions if it would help.

asumaran commented:
@dechov Thanks for your response. I think I figured out the issue.

I've created an issue with the following custom attributes 

<img width="315" alt="Screen Shot 2021-05-05 at 2 44 15 AM" src="https://user-images.githubusercontent.com/1025173/117110485-cab5da00-ad4b-11eb-81ac-95eb89c9df2d.png">

And the following variations where price is set as follows: 5, 10, 15, 20, 25.

<img width="272.5" alt="Screen Shot 2021-05-05 at 2 44 50 AM" src="https://user-images.githubusercontent.com/1025173/117110536-e02b0400-ad4b-11eb-9cec-7e805e8ce147.png">

To select a variation via GET we need to use the attribute's name exactly as is saved in the DB:

```
mysql> SELECT meta_key, meta_value FROM `wp_postmeta` WHERE `post_id` = 404 AND `meta_key` LIKE 'attribute_%';
+---------------------+------------+
| meta_key            | meta_value |
+---------------------+------------+
| attribute_la-edad   | Muy Joven  |
| attribute_el-tamano | Muy Grande |
+---------------------+------------+
2 rows in set (0.00 sec)
```

So, in this case the query string we need to pass via GET will be like `?attribute_la-edad=Muy Joven`. Notice we are passing the value exactly as it is in the DB.

On the other hand, variations created with predefined attributes the attribute name has a different prefix and also its value is sanitized.

```
mysql> SELECT meta_key, meta_value FROM `wp_postmeta` WHERE `post_id` = 368 AND `meta_key` LIKE 'attribute_%';
+--------------------+-------------+
| meta_key           | meta_value  |
+--------------------+-------------+
| attribute_pa_size  | extra-large |
| attribute_pa_color | red         |
+--------------------+-------------+
2 rows in set (0.00 sec)
```

In this case if we want to select an existing variation via GET we need to use the following format: `?attribute_pa_size=extra-large`.

Let me know if this makes sense to you. I think everything is working as expected now. Otherwise let's do a call to quickly fix anything you might find 👍

dechov commented:
@asumaran Thanks for testing and elaborating on those predefined vs. custom attribute cases – I hadn't noticed the different behavior, or that bug.

The case I was describing related not to query parameters but to the default form values: when I had a custom product attribute with a name like "Size" and set its default value, the product page would load with the correct matching variation but the payment request flow initially wouldn't.

dechov commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -312,7 +312,13 @@ public function get_product_data() {
+
+			foreach ( $variation_attributes as $attribute_name => $attribute_values ) {
+				$attribute_key                = 'attribute_' . sanitize_title( $attribute_name );
+				$attributes[ $attribute_key ] = isset( $_GET[ $attribute_key ] ) ? wc_clean( wp_unslash( $_GET[ $attribute_key ] ) ) : $product->get_variation_default_attribute( $attribute_name );
```
This line is pretty long (though not at all the longest in the file) so I'd vote to split the ternary operands into separate lines, and/or add a code comment with a more concise description of the logic here.

asumaran replied:
done https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1533/commits/4bbb710ab09d7f93e63c76c3ecffafff320a8841

dechov reviewed (approved):
Apologies, didn't realize the latest commit did in fact sanitize the attribute name – looks good & tests well for me now. Tested with several different attributes and query param combinations. Left one minor comment but otherwise good to go :shipit:

asumaran committed:
Split ternary operator into multiple lines

asumaran commented:
Thanks for review @dechov
